### ISSUE TITLE
Function Arguments of named tuples can't infer param types despite explicit type

### ISSUE THREAD

**Issue by bishabosha** [scala/scala3#20456]
## Compiler version

3.5.0-RC1

## Minimized code

```Scala
import scala.language.experimental.namedTuples

val m: (foo: String => Int, bar: String => Int) = (foo = _.length, bar = _.length) // error // error
val n: (String => Int, String => Int) = (_.length, _.length)
```

## Output

```scala
[error] foo.scala:30:60
[error] Missing parameter type
[error] 
[error] I could not infer the type of the parameter _$5
[error] in expanded function:
[error]   _$5 => _$5.length
[error] Expected type for the whole anonymous function:
[error]   Any
[error]   val m: (foo: String => Int, bar: String => Int) = (foo = _.length, bar = _.length)
[error]                                                            ^
[error] foo.scala:30:76
[error] Missing parameter type
[error] 
[error] I could not infer the type of the parameter _$6
[error] in expanded function:
[error]   _$6 => _$6.length
[error] Expected type for the whole anonymous function:
[error]   Any
[error]   val m: (foo: String => Int, bar: String => Int) = (foo = _.length, bar = _.length)
[error]                                                                            ^
```

## Expectation

correctly infer the arguments, as with normal tuples.

**Use Case**

I would like to be able to construct programatically a type for a named tuple where all its fields are function types, and then the user should be able to construct this object without explicit parameter types

**Comment by bishabosha** [comment]
This looks like its because of the desugaring - i.e
```scala
(foo =  _.length) : (foo: String => Int)
```

becomes

```scala
NamedTuple.withNames(Tuple1(_.length))[Tuple1["foo"]] : (foo: String => Int)
//                          ^^^^^^^^
//                          nested lambda with no expected type
```

it might be worth reconsidering this desugaring, or perhaps tunnelling through any expected type to the argument of `withNames`

**Comment by odersky** [comment]
Named tuples are constructed with `NamedTuple.withNames`:

```scala
  extension [V <: Tuple](x: V)
    inline def withNames[N <: Tuple]: NamedTuple[N, V] = x
```
So the call looks like 
```scala
NamedTuple.withNames[V](tuple)[N]
```
where typically `V` is inferred but `N` is explicit. It's important to have two type parameter lists because the `V` is usually inferred but the `N` needs to be given explicitly. But they don't nesessarily need to come in that order. One could also imagine to build named tuples using a function like this
```scala
def build[N <: Tuple]()[V <: Tuple](x: V)
```
Maybe that would fix the problem. I think it might be worth trying out, but I don't have the bandwidth to do it myself right now.

**Comment by bishabosha** [comment]
This solution seems to work without needing clause interleaving, and leaves no trace after inlining:
```scala
object NamedTupleBuilder {
  opaque type Builder[N <: Tuple] = Unit
  def apply[N <: Tuple]: Builder[N] = ()
}
extension [N <: Tuple](inline b: NamedTupleBuilder.Builder[N])
  transparent inline def build[V <: Tuple](v: V): NamedTuple[N, V] = v.withNames[N]
```

**Comment by bishabosha** [comment]
I have implemented a fix in #20497
